### PR TITLE
feat(appmesh): add IAM grants for StreamAggregatedResources

### LINKED ISSUE #11639: [appmesh] Add IAM Grants to Resources
Per the [CDK Design Guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md#grants) App Mesh should have grants on a per resource basis.

### Use Case

Granting permissions to other services to call App Mesh is an important part of stabilizing the App Mesh L2 APIs

### Proposed Solution

For each resource, we should implement a grant for `Describe`, `Delete`, `Update`, `TagResource`, and `UntagResource`. There should also be high level `ListAll` and `DescribeAll` permissions as described in the [CDK Design Guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md#grants).

For Virtual Nodes and Virtual Gateways, we should implement a grant for `StreamAggregatedResources` 

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by alexbrjo**
This adds the IAM grant `grantStreamAggregatedResources` to VirtualNodes and VirtualGateways. Example below

```ts
const gateway = new appmesh.VirtualGateway(stack, 'testGateway', { mesh: mesh });
const envoyUser = new iam.User(stack, 'envoyUser');

/**
 * This will grant `grantStreamAggregatedResources` ONLY for this gateway.
 */
gateway.grantStreamAggregatedResources(envoyUser)
``` 

resolves #11639

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by skinny85 — COMMENTED** [review]
Thanks for submitting @alexbrjo. Before I dive into the details of this, a general comment/question.

In general, our `grant` methods try to be intent-based, with names like `grantRead()`, `grantWrite()`, `grantConnect()`, etc. I have to say, I'm not a huge fan of methods just called `grantAll`.

Is there any way we can support more fine-grained `grant()` methods? Can we divide the use-cases these resources take part in and minimize the permissions assigned to each principal, instead of just granting always all permissions in all cases?

Thanks,
Adam

**Comment by alexbrjo** [comment]
I choose to use `grantAll` instead of `grantRead`/`grantWrite` because: (1) I didn't think it'd be that useful to grant an entity read permissions for a single resource in a mesh. (2) If you're giving an entity permission to write or delete a resource, it seems unlikely that you would want to deny that entity read permissions. However, I agree with everything you wrote and since there are already conventions it's best to follow them.

I'd like your feedback on a hierarchical granting pattern that I think would be useful for App Mesh. Mesh, Virtual Router and Virtual Gateway would extra`grant[Read|Write]AllSubresources`. Our ARNs are hierarchical so accomplishing this could be pretty easy. The only downside is that it would be granting permissions to resources that do not exist (ex. granting read access to all resources in a mesh). Is this acceptable for CDK grants? 

For reference here is the hierarchy of App Mesh resources:

- mesh
  - Virtual Service
  - Virtual Node
  - Virtual Router
    - Route 
  - Virtual Gateway
    - Gateway Route

**Comment by skinny85** [comment]
For the hierarchical resources, I would do it like this:

1. Each resource grants permissions to all of its subresources when `grant*()` is called on it. So, if I call `mesh.grantRead()`, I should get read permissions to the Mesh itself, and all of its subresources.
2. Every resource has their own `grant()` methods, if I want to be more fine-grained. For example, if I want to give someone read access to a single Virtual Node - not the entire Mesh - there should be a `grantRead()` on the Virtual Node that will allow me to do that.

Does this make sense?

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/mesh.ts:165]: I wonder - can we separate the permissions for these that are for the Mesh's ARN, and for those that require Mesh ARN + `'/*'`?

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/mesh.ts:168]: No need for the empty line 😀

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:40]: No need for the empty line 😊

File: packages/@aws-cdk/aws-appmesh/lib/route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route.ts:105]: But you _do_ need an empty line before this one 😜

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:169]: Same question here (can we divide the permissions between the 2 resources that are needed for them?).

File: packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts:46]: Unnecessary empty line 😜

File: packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts:133]: Same question here (can we split the permissions on the 2 resources?).

File: packages/@aws-cdk/aws-appmesh/test/test.gateway-route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.gateway-route.ts:226]: Empty line between the tests pliz, thanks!

Same comment in the other test files.

File: packages/@aws-cdk/aws-appmesh/test/test.mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:366]: If you could indent this to save some lines:

```suggestion
              {
                'Fn::Join': ['', [
                  { Ref: 'meshACDFE68E' },
                  '/*',
                ]],
              },
```

Same comment for the other tests.

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great @alexbrjo!

Some questions about the exact resources used in the IAM policies, and some nitpicking about empty lines and indentation 😉.

**Comment by skinny85** [comment]
One more comment.

I see that some actions are duplicated between the hierarchies of the resources that you mentioned in https://github.com/aws/aws-cdk/pull/13596#issuecomment-801209614.

For example, for `VirtualGateway` and `GatewayRoute`:

```ts
  public grantRead(identity: iam.IGrantable): iam.Grant {
    return this.grant(identity,
      'appmesh:DescribeVirtualGateway',
      'appmesh:ListVirtualGateway',
      'appmesh:DescribeGatewayRoute',
      'appmesh:ListGatewayRoute',
    );
  }

...

  public grantRead(identity: iam.IGrantable): iam.Grant {
    return this.grant(identity,
      'appmesh:DescribeGatewayRoute',
      'appmesh:ListGatewayRoute',
    );
  }
```

Any chance we could get rid of this duplication?

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (alexbrjo) [reply packages/@aws-cdk/aws-appmesh/lib/mesh.ts:165]: Currently it looks like grants don't support multiple policy statements.

**Comment by alexbrjo** [comment]
> Any chance we could get rid of this duplication?

If grants supported multiple policy statements I think we could recursively build the policy statements pretty easily. We could remove the duplication by writing additional wrapper methods, but IMO the extra code isn't worth it. 

Are you worried about us forgetting to update it in all spots in the future?

**Comment by alexbrjo** [comment]
> Any chance we could get rid of this duplication?

If grants supported multiple policy statements I think we could recursively build the policy statements pretty easily. We could remove the duplication by writing additional wrapper methods, but IMO the extra code isn't worth it. 

Are you worried about us forgetting to update it in all spots in the future?

**Comment by skinny85** [comment]
I'm worried because of all of the usual problems code duplication causes. Us forgetting to update them in 2 places is one of them, yes 🙂.

**Comment by alexbrjo** [comment]
I'm waiting for feedback from my team about the recursive grants. Assuming they approve of it I'll add in wrapper methods to collection actions.

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/mesh.ts]: How about we call `Grant.addToPrincipal()` twice - once with actions on the Mesh, and `resourceArns` equal to `[this.meshArn]`, and then a second time, with the sub-resource actions, and `resourceArns` equal to ``[`${this.meshArn}/*`]``, and return the result of that first call?

**Comment by alexbrjo** [comment]
We're going to stick with non-recursive grants for now. Updating this PR

File: packages/@aws-cdk/aws-appmesh/README.md
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/README.md:625]: We should include `ListTagsForResource` in these Read Policies as well.\

**Comment by alexbrjo** [comment]
Hey @skinny85, this one is ready for review again. I've updated this PR to grant permissions non-recursively. Thanks!

File: packages/@aws-cdk/aws-appmesh/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:597]: > `grantRead` grants the user read access to the resource all of it's subresources.

Is this true still?

File: packages/@aws-cdk/aws-appmesh/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:598]: > And similarly `grantWrite` grants the user write access to the resource to all subresources.

And similarly here 🙂.

File: packages/@aws-cdk/aws-appmesh/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:615]: I feel like we're just asking for this to become stale ;p. Maybe we don't have to include the entire list here...?

File: packages/@aws-cdk/aws-appmesh/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:625]: Same comment here (skip the exact list?).

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: I have to say, I'm a little surprised by these permissions here.

This is the `GatewayRoute` resource itself. Why would be allow creating and deleting it to other resources? I understand that the IAM principal performing the CloudFormation deployment needs these permissions... but why would a different principal from the same Stack as this Gateway Route be allowed to delete it?

@dfezzie  What's the use case for having these permissions?

File: packages/@aws-cdk/aws-appmesh/lib/mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/mesh.ts:128]: Same questions here as for `GatewayRoute` - what would necessitate a resource having these permissions?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:47]: If the JSDocs are the same as the implemented property / method, there's no reason to repeat them for subclasses.

Same comment everywhere below.

File: packages/@aws-cdk/aws-appmesh/lib/route.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route.ts:81]: Same comment as in `GatewayRoute` (no reason to repeat JSDocs for overridden members).

Notice for example that in `MeshBase`, the `grant*()` methods don't have JSDocs repeated.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:143]: Same questions here as for `GatewayRoute` - what would necessitate a resource having these permissions?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:155]: Shouldn't this be

```suggestion
    return this.grant(identity, 'appmesh:StreamAggregatedResources');
```

?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:159]: Shouldn't this be

```suggestion
    return this.grant(identity, 'appmesh:StreamAggregatedResources');
```

?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:147]: Same questions here as for `GatewayRoute` - what would necessitate a resource having these permissions?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts:117]: Same questions here as for `GatewayRoute` - what would necessitate a resource having these permissions?

File: packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:104]: Same questions here as for `GatewayRoute` - what would necessitate a resource having these permissions?

File: packages/@aws-cdk/aws-appmesh/test/test.mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:319]: Missing empty line before this line.

File: packages/@aws-cdk/aws-appmesh/test/test.mesh.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.mesh.ts:351]: Missing empty line before this line.

**Review by skinny85 — DISMISSED** [review]
In general looks fine - there are some needlessly repeated JSDocs in the base classes, and a few places where a helper method has not been used.

The bigger problem is with the permissions given by the `grantWrite()` methods, and I have to say I don't understand what they are supposed to represent. To me they look like permissions used to _deploy_ a given resource - not permissions used by other resources when interacting with the given resource.

@dfezzie can you explain what is the use case for other resources using `grantWrite()` methods on the AppMesh resources?

File: packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: I'm a little confused on what you mean. Can you explain what the purpose of IAM grants is in general? I thought we needed to provide these base permission groups for our resources to be considered feature complete for CDK. How do other teams typically implement IAM grants?

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: Yeah, I think there's some confusion here 🙂. Let's look at the existing `grant*()` APIs as a reference, maybe. that will clears things up.

Let's take S3 as an example. when you call `bucket.grantWrite(principal)`, it grants the permissions `s3:PutObject` and `s3:Abort*` on that specific Bucket. These are used by resources that write to the Bucket.

But note that it _doesn't_ grant `s3:CreateBucket`, nor `s3:DeleteBucket`!

1. The Bucket obviously already exists, so there's no point in granting `s3:CreateBucket` on its ARN.
2. The resource should write things to the Bucket -it should _not_ have permissions to delete the entire Bucket! Only the deployment Role needs those permissions.

> I thought we needed to provide these base permission groups for our resources to be considered feature complete for CDK. How do other teams typically implement IAM grants?

No. There is no such requirement. There are many, many resources in the CDK that do _not_ have `grant*()` methods, because they don't make sense for that resources (for example, what would `grantRead()` mean for a Load Balancer?).

----------------

Does this make sense @dfezzie?

Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: Yeah, that does make sense. 

So, with that in mind, I think the _most_ applicable IAM Grant for us is `grantStreamAggregatedResources`. This is needed for granting whatever compute your mesh services are running on to fetch the config from our Envoy Management Service. 

I could also imagine a scenario where you would need to grant high level admin permissions to manage cluster resources to some role. For example, we have an `App Mesh Controller` that manages the lifecycle of App Mesh objects in K8s. The master node there would need those permissions, but that doesn't work with the current approach in this PR.

@alexbrjo what are your thoughts on this?

Comment (alexbrjo) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: I included create and delete simply because they are "write" actions.

An update (which can break routing configuration) is on the same level of access as a delete or a create (which in this case would be a re-creation of a previously existing resource). The only benefit of excluding them is if the customer use case is to grant the ability to modify a resource but safe guard against a deletion, but this is speculation.

Comment (alexbrjo) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: Thanks for explaining this, Adam. I thought we wanted to include all actions for each resource. In this case I don't think it makes sense to include control plane actions unless we're doing a recursive `grantAdmin` control of the mesh for k8s controller. I think we could go with only `grantStreamAggregatedResources`.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/gateway-route.ts:82]: Sounds good @alexbrjo - let's go with only `grantStreamAggregatedResources()`.

**Comment by skinny85** [comment]
@alexbrjo is this ready for another round of reviews?

**Review by skinny85 — DISMISSED** [review]
Requesting changes to clear this from my ToDo list.

**Comment by alexbrjo** [comment]
@skinny85 Yep, this PR is now just the grant for `StreamAggregatedResources`

**Review by skinny85 — APPROVED** [review]
Looks great, thanks @alexbrjo!

**Pull request merged**
